File: android/app/src/main/java/dev/carraes/lam/articles/Article.kt
package dev.carraes.lam.articles

import kotlinx.serialization.SerialName
import kotlinx.serialization.Serializable
import kotlinx.serialization.json.Json
import kotlinx.serialization.json.JsonElement
import java.security.MessageDigest

internal val articleJson = Json { ignoreUnknownKeys = true; encodeDefaults = true }

@Serializable
data class ArticleAsset(val path: String, @SerialName("media_type") val mediaType: String,
    val size: Long, val sha256: String, val disposition: String) {
    val inlineImage: Boolean get() = disposition == "inline" && mediaType in setOf("image/png", "image/jpeg", "image/webp")
}

@Serializable
data class Article(val id: String, val title: String, val summary: String, val name: String,
    @SerialName("source_host") val sourceHost: String, @SerialName("source_project") val sourceProject: String,
    @SerialName("created_at") val createdAt: String, @SerialName("read_at") val readAt: String?,
    val version: Long, val assets: List<ArticleAsset>)

@Serializable
data class ArticlePage(val items: List<Article>, @SerialName("next_cursor") val nextCursor: String?)

@Serializable
data class ArticleContent(val article: Article, val parts: List<JsonElement>)

data class ArticleSession(val account: String, val epoch: String, val generation: Long = 0)
data class LoadedArticle(val article: Article, val html: String, val session: ArticleSession, val cached: Boolean = false)
enum class ReadResult { SAVED, CONFLICT, FAILED }

internal fun sha256(bytes: ByteArray): String = MessageDigest.getInstance("SHA-256").digest(bytes).joinToString("") { "%02x".format(it) }

File: worker/src/domain/Article.ts
import { Data, Schema } from "effect";

export const Asset = Schema.Struct({
  path: Schema.String,
  media_type: Schema.String,
  size: Schema.Number,
  sha256: Schema.String,
  disposition: Schema.Literal("inline", "attachment"),
});
export type Asset = typeof Asset.Type;
export const ArticleDraft = Schema.Struct({
  title: Schema.String,
  summary: Schema.String,
  name: Schema.String,
  source_host: Schema.String,
  source_project: Schema.String,
  silent: Schema.Boolean,
  assets: Schema.Array(Asset),
});
export type ArticleDraft = typeof ArticleDraft.Type;
export interface Article {
  id: string;
  title: string;
  summary: string;
  name: string;
  source_host: string;
  source_project: string;
  created_at: string;
  read_at: string | null;
  version: number;
  assets: readonly Asset[];
}
export interface ArticlePage { items: Article[]; next_cursor: string | null }
export const ReadUpdate = Schema.Struct({ read: Schema.Boolean, version: Schema.Int.pipe(Schema.nonNegative()) });
export type ReadUpdate = typeof ReadUpdate.Type;
export class ArticleStorageError extends Data.TaggedError("ArticleStorageError")<{ cause: unknown }> {}
